### PR TITLE
Remove nProtocolVersion and add mode/type fields to DIP3

### PR DESCRIPTION
**Pull request by codablock**
See individual commits.

These were last-minute changes in the DIP3 docs and never updated in-code.

### CODE REVIEW

**Comment by codablock** [comment]
@UdjinM6 asked for a tl;dr description of why we initially decided to remove this from DIP3. Here is a copy of the reply on Slack:

```
So...we had a long discussion back then. The reasons we removed nProtocolVersion in the end were:
1. The version is defined by the version of the controller wallet running on your local machine and not by the dashd version running on the MN. This means we'll end up with MNs who accidentally issue ProUpServTx with mismatching wallet versions. So MNs will be registered as latest version, but actually still running the old version. User won't notice this easily.
2. People might actually do this on purpose...incentives to do so are high (no time to upgrade? just issue "protx update_servive") because MNs are not punished for this.
3. Looking at where the proto version is used, it seems to be not required when switching to deterministic MNs. It's mostly used for payment logic (which is completely different in DIP3) and for private send. for private send, it's only used to filter for minimum proto version. that case can be easily reimplemented by using the proto version of the connection and then just disconnecting when connected to an old MN
4. IMHO, we should do upgrade logic in a way that doesn't require the MN proto version. We can do BIP9, the network proto version, sporks and so on for upgrades

For me, most important thing is 1. and 2. as they make the MN proto version completely unreliable. In the past, people had to modify source code to fake a wrong version. Now they just have to call a RPC.
If we ever want to reimplement something to enforce minimum dashd version, it should be part of PoSe
```

**Comment by codablock** [comment]
@UdjinM6 asked for a tl;dr description of why we initially decided to remove this from DIP3. Here is a copy of the reply on Slack:

```
So...we had a long discussion back then. The reasons we removed nProtocolVersion in the end were:
1. The version is defined by the version of the controller wallet running on your local machine and not by the dashd version running on the MN. This means we'll end up with MNs who accidentally issue ProUpServTx with mismatching wallet versions. So MNs will be registered as latest version, but actually still running the old version. User won't notice this easily.
2. People might actually do this on purpose...incentives to do so are high (no time to upgrade? just issue "protx update_servive") because MNs are not punished for this.
3. Looking at where the proto version is used, it seems to be not required when switching to deterministic MNs. It's mostly used for payment logic (which is completely different in DIP3) and for private send. for private send, it's only used to filter for minimum proto version. that case can be easily reimplemented by using the proto version of the connection and then just disconnecting when connected to an old MN
4. IMHO, we should do upgrade logic in a way that doesn't require the MN proto version. We can do BIP9, the network proto version, sporks and so on for upgrades

For me, most important thing is 1. and 2. as they make the MN proto version completely unreliable. In the past, people had to modify source code to fake a wrong version. Now they just have to call a RPC.
If we ever want to reimplement something to enforce minimum dashd version, it should be part of PoSe
```

File: src/version.h [hunk 1]
```diff
@@ -44,10 +44,4 @@ static const int DIP0001_PROTOCOL_VERSION = 70208;
 //! short-id-based block download starts with this version
 static const int SHORT_IDS_BLOCKS_VERSION = 70209;
 
-//! minimum ProTx proto version
-static const int MIN_PROTX_PROTO_VERSION = 70211;
```
Comment (UdjinM6) [inline src/version.h]: I would change this to
```
//! Deterministic Masternodes were introduced in this version
static const int DMN_PROTO_VERSION = 70211;
```
and ...

File: src/masternode.cpp [hunk 2]
```diff
@@ -55,7 +55,7 @@ CMasternode::CMasternode(const CMasternodeBroadcast& mnb) :
 {}
 
 CMasternode::CMasternode(const uint256 &proTxHash, const CDeterministicMNCPtr& dmn) :
-    masternode_info_t{ MASTERNODE_ENABLED, dmn->pdmnState->nProtocolVersion, GetAdjustedTime(),
+    masternode_info_t{ MASTERNODE_ENABLED, PROTOCOL_VERSION, GetAdjustedTime(),
```
Comment (UdjinM6) [inline src/masternode.cpp]: ... use it here (i.e.  s/PROTOCOL_VERSION/DMN_PROTO_VERSION/) ...

File: src/masternodeman.cpp [hunk 3]
```diff
@@ -391,7 +391,7 @@ void CMasternodeMan::AddDeterministicMasternodes()
             mn->blsPubKeyOperator = dmn->pdmnState->pubKeyOperator;
             mn->keyIDVoting = dmn->pdmnState->keyIDVoting;
             mn->addr = dmn->pdmnState->addr;
-            mn->nProtocolVersion = dmn->pdmnState->nProtocolVersion;
+            mn->nProtocolVersion = PROTOCOL_VERSION;
```
Comment (UdjinM6) [inline src/masternodeman.cpp]: ... and here

**Comment by codablock** [comment]
@UdjinM6 What is the reasoning behind setting proto version to a fixed value? My idea behind simply using PROTOCOL_VERSION was that this is all happening on the local MN and thus doesn't matter which version we actually use. If we have DMN_PROTO_VERSION, how would it help in upgrades? If we try to filter by it in the legacy code, it will always either return all MNs or none, depending on if we did a bump or not.

**Comment by UdjinM6** [comment]
Well, imo using local proto version as a MN version is simply wrong :) Using some static version is also not quite right but at least we _know_ it should be 70211 for DMNs and it kind of resonates with the idea behind `version.h` (keeping track of versions when some specific features were introduced), so why not?

**Comment by codablock** [comment]
Convinced :) Updated and pushed.

**Pull request merged**